### ISSUE TITLE
Optic methods

### ISSUE THREAD

**Issue by Siegrift** [TSOptics/optics#4]
Hi, I came across your article https://unfocused.blog/posts/comparing-optics-with-jotai and I really like the concept.

I haven't spend much time reading the internals, but I am not sure I understand the motivation of the optic methods (e.g. `findFirst `, `refine`, `max`...). This seems to me like the library is trying to re-implement lodash (or similar) which I believe should not be its concern. This might also be bad for codesplitting.

Are there some advantages of implementing this inside the library and not in user-land?

**Comment by VincentQuillien** [comment]
Hi !

These methods return a new optic focused on a value, and not the value itself.
For example with an optic focused on an array, `findFirst` derives **a new optic** focused on the first element matching the predicate, and not the element itself.
So they don't serve the same purpose as the similarly named methods from the standard library or the util functions from lodash.

Instead of making them methods of Optic I could have shipped them as functions that we pass to `derive`:
```ts
import { findFirst } from "@optics/react"

const onFirstEven = onNumbers.derive(findFirst(x => x % 2 === 0));
```

I has the advantage of being tree-shakable, as well as making the built-in functions compose the same way as the functions built in user-land.
The downsides is that it's more verbose and we loose auto-completion but it might be worth the trade-off.
WDYT ?

**Comment by Siegrift** [comment]
> These methods return a new optic focused on a value, and not the value itself.
For example with an optic focused on an array, findFirst derives a new optic focused on the first element matching the predicate, and not the element itself.

Yeah, but for example in order to implement `findFirst` optic derivation, you need to do `array.findFirst`. If you want to find first value in an object (record) you need to go for lodash (or some custom implementation). To me, it feels like the `derive` function could be enough in its essence (and the library doesn't need to define these derivations). 

In my experience, what I always needed/wanted was for the state to compose, such as slices in redux toolkit or in this library case, the natural lenses composition. I wouldn't want to derive new optics that transform the original one because essentially you're creating two way data sync between two pieces of state which I dislike. So I'd mostly use the `.derive` method to create more focused optics (e.g. array element, record value) and for that I'd just need which index/key to select/update.

> I has the advantage of being tree-shakable, as well as making the built-in functions compose the same way as the functions built in user-land.
The downsides is that it's more verbose and we loose auto-completion but it might be worth the trade-off.

Yeah, that's sounds good to me as well.

**Comment by VincentQuillien** [comment]
I've just released version 0.3.0 that replaces methods by functions called combinators, that we can call inside the `derive` method to get new optics: https://www.optics.page/docs/Introduction/core-concepts/#combinators

There's no difference between a built-in combinator and a custom one, and they are straight-forward to create as it's just a function returning a `get` and a `set` function: https://github.com/TSOptics/optics/blob/master/packages/core/src/combinators/find.ts
